### PR TITLE
fix: [CDS-44543]: Email step

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)

### CODE REVIEW

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 1]
```diff
@@ -7,51 +7,188 @@
...
-      Ambiance ambiance, StepElementParameters stepParameters, StepInputPackage inputPackage) {
-    return null;
-  }
+  public StepResponse executeSync(Ambiance ambiance, StepElementParameters stepParameters,
+      StepInputPackage inputPackage, PassThroughData passThroughData) {
+    long startTime = System.currentTimeMillis();
+    NGLogCallback logCallback = new NGLogCallback(logStreamingStepClientFactory, ambiance, null, true);
+    EmailStepParameters emailStepParameters = (EmailStepParameters) stepParameters.getSpec();
+    String toMail = emailStepParameters.to.getValue();
+    String ccMail = emailStepParameters.cc.getValue();
```
Comment (shivamnegi94) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:75]: Why ccMail and toMail are assigned to getValue()?

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java (hunk 1 above)
Comment (vivekkarnatiharness) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:75]: to present in emailparameters are parameter fields ,to get the string value,we have to use .getvalue()!

**Review by shivamnegi94 — APPROVED** [review]

**Review by shivamnegi94 — DISMISSED** [review]

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 2]
```diff
@@ -7,51 +7,188 @@
...
+import org.apache.commons.lang3.StringUtils;
+import retrofit2.Response;
 
 @Slf4j
 @OwnedBy(HarnessTeam.CDC)
-public class EmailStep extends TaskExecutableWithRollback<EmailStepResponse> {
+public class EmailStep implements SyncExecutable<StepElementParameters> {
+  @Inject private NotificationClient notificationClient;
   public static final StepType STEP_TYPE = StepSpecTypeConstants.EMAIL_STEP_TYPE;
+  private static final String INFRASTRUCTURE_COMMAND_UNIT = "Execute";
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:58]: Please check if we can make do without passing command unit altogether.

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 3]
```diff
@@ -7,51 +7,188 @@
...
+            .unitProgressList(Collections.singletonList(UnitProgress.newBuilder()
+                                                            .setUnitName(INFRASTRUCTURE_COMMAND_UNIT)
+                                                            .setStatus(UnitStatus.FAILURE)
+                                                            .setStartTime(startTime)
+                                                            .setEndTime(System.currentTimeMillis())
+                                                            .build()))
+            .build();
+      }
+
+    } catch (SocketTimeoutException e) {
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:130]: Let's not handle SockerTimeoutException separately. There is very little value that we are adding. Just catching IOException will do.

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 4]
```diff
@@ -7,51 +7,188 @@
...
+      logCallback.saveExecutionLog(String.format("Not able to send emails,please check SMTP configuration"),
+          LogLevel.INFO, CommandExecutionStatus.FAILURE);
+      log.error("Not able to send emails", e.getCause());
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:133]: `log.error("Not able to send emails", e);`

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 5]
```diff
@@ -7,51 +7,188 @@
...
+                  .build())
+          .unitProgressList(Collections.singletonList(UnitProgress.newBuilder()
+                                                          .setUnitName(INFRASTRUCTURE_COMMAND_UNIT)
+                                                          .setStatus(UnitStatus.FAILURE)
+                                                          .setStartTime(startTime)
+                                                          .setEndTime(System.currentTimeMillis())
+                                                          .build()))
+          .build();
+    } catch (IOException e) {
+      logCallback.saveExecutionLog(String.format("Not able to send email due to a internal error"), LogLevel.INFO,
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:156]: Be a bit more verbose here. 
```
Failed to send the email. The reasons could be -
- Possible errors on the Harness side
- Delegate unable to reach custom SMTP server(if configured)
- The SMTP server may not be setup correctly(if configured)
```
You do not need to use the message as is, but please try to highlight all possible reasons why we could send the email. The reason for Sockertimeout would be the same. 

Let's merge IO & Sockertimeout error

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java (hunk 2 above)
Comment (vivekkarnatiharness) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:58]: Yeah working!

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 6]
```diff
@@ -7,51 +7,166 @@
...
+      }
+
+    } catch (IOException e) {
+      logCallback.saveExecutionLog(String.format("Failed to send the email. The reasons could be -\n"
+                                       + "- Harness platform service might be not working\n"
+                                       + "- Delegate unable to reach custom SMTP server(if configured)\n"
+                                       + "- The SMTP server may not be setup correctly(if configured)"),
+          LogLevel.INFO, CommandExecutionStatus.FAILURE);
+      log.error("Not able to send emails", e);
+      return StepResponse.builder()
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:139]: We should not have any outcomes for failed steps.

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 7]
```diff
@@ -7,51 +7,166 @@
...
+            .build();
+      }
+      if (response.body().getStatus() == io.harness.ng.core.Status.SUCCESS
+          && StringUtils.isNotBlank(response.body().getData().getErrorMessage())) {
+        logCallback.saveExecutionLog(String.format(response.body().getData().getErrorMessage()));
+      }
+
+    } catch (IOException e) {
+      logCallback.saveExecutionLog(String.format("Failed to send the email. The reasons could be -\n"
+                                       + "- Harness platform service might be not working\n"
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:134]: Customer does not know what all microservices we have & they do not need to know what microservices we do have. 
rephrase the message as `Something went wrong on Harness's end.`
Put this as the last option & not the first option.

**Review by puthrayaharness — APPROVED** [review]

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java [hunk 8]
```diff
@@ -7,51 +7,166 @@
...
 
   @Override
-  public TaskRequest obtainTask(
-      Ambiance ambiance, StepElementParameters stepParameters, StepInputPackage inputPackage) {
-    return null;
-  }
+  public StepResponse executeSync(Ambiance ambiance, StepElementParameters stepParameters,
+      StepInputPackage inputPackage, PassThroughData passThroughData) {
+    long startTime = System.currentTimeMillis();
+    NGLogCallback logCallback = new NGLogCallback(logStreamingStepClientFactory, ambiance, null, true);
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/email/EmailStep.java:74]: Where are we closing the logCallback? Can we close the logCallback if the step is aborted?

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger ti1

**Comment by vivekkarnatiharness** [comment]
trigger ti2

**Comment by vivekkarnatiharness** [comment]
trigger ti4

**Comment by vivekkarnatiharness** [comment]
trigger compile

**Comment by vivekkarnatiharness** [comment]
trigger ft1

**Review by puthrayaharness — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 4 of 6 files at r2, 4 of 5 files at r4, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by puthrayaharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by naman-goenka — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by naman-goenka — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by naman-goenka — APPROVED** [review]
> Reviewed 3 of 8 files at r1, 3 of 6 files at r2, 4 of 5 files at r4, 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by shivamnegi94 — COMMENTED** [review]
> Reviewed 3 of 8 files at r1, 3 of 6 files at r2, 4 of 5 files at r4, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: 1. unresolved discussions.

**Review by shivamnegi94 — APPROVED** [review]

**Review by shivamnegi94 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39174)* status: :shipit: complete! ✓

**Pull request merged**